File: src/retar/finalizer.ts
import { systemRoles } from '../types/index.ts'
import type { VillageStatus, SystemRole, Seat, Day, SeatStatus } from '../types/index.ts'
import type { Possibilities } from './possibilities.ts'
import type { AnalyzeContext } from './roleTesters.ts'
import { solvePossibilities } from './solver.ts'
import { dumpFinalizePre } from './dump.ts'
import {
  singleRoleByTrait,
  singleRoleBySeerResult,
  countByTraitIn,
  countBySeerResultIn,
  rolesWithTraitIn,
} from './role-sets.ts'

// hot path で繰り返し呼ばれるため module-level 解決.
const foxRole = singleRoleByTrait('passive', 'die-when-divined')
const wolfRole = singleRoleBySeerResult('wolf')

/**
 * seat が夜 `day` の時点で行動可能 (alive) かどうか.
 * 夜 D の causeOfDeath === 'night_kill' は「その夜に襲撃で死亡」 = その夜の行動は可能, とみなす
 * (verifyDivineAbility の seer maxActiveDay と同じ流儀).
 * execution は昼の処刑なので diedDay === D でも夜 D は alive ではない.
 */
function isAliveAtNight(status: SeatStatus, day: Day): boolean {
  if (status.surviving) return true
  if (status.diedDay! > day) return true
  if (status.diedDay! === day && status.causeOfDeath === 'night_kill') return true
  return false
}

/**
 * 死体数の検証。各日の夜死体数が仮説と整合するか確認する。
 * @returns true: 検証パス、false: 仮説を棄却
 * 副作用: context.requireOneOf に背徳後追い制約を追加する場合がある
 */
// 読み取り専用版: 夜死者数の矛盾チェックのみ。context を変更しない。
export function checkDeathCounts(
  context: AnalyzeContext,
  vs: VillageStatus,
  nightKillsByDay: Map<Day, Seat[]>,
  setup: Map<SystemRole, number>,
): boolean {
  const immoralists = countByTraitIn(setup, 'reactive', 'follow-fox-death')
  const guardRoles = rolesWithTraitIn(setup, 'action', 'guard')
  DAY:
  for ( const [day, killed] of nightKillsByDay.entries() ) {
    if ( vs.day <= day ) continue DAY
    const addCount = context.deathChronicle.add[day]
    let expected = 1
    if ( addCount ) expected += addCount
    const actual = killed.length
    if ( actual === expected ) continue DAY
    if ( expected + immoralists < actual ) {
      return false
    }
    else if ( actual < expected - 1 ) {
      return false
    }
    else if ( expected < actual && actual <= expected + immoralists ) {
      const hamsterDiedThisNight = context.hamstersKilledByDivine.some(h => h.day === day)
      if ( hamsterDiedThisNight ) {
        // mut版では requireOneOf に制約を追加する。読み取り専用版では矛盾なしとして通す
        continue DAY
      }
    }
    if ( actual < expected ) {
      // peace night (actual < expected) を成立させる説明:
      //   (A) 夜 day に alive な妖狐がいる (狼襲撃先 = 妖狐 → 襲撃免疫で死体なし)
      //   (B) 夜 day に alive な狩人がいる (護衛成功)
      // 「狐 or 狩人」 を possibilities で見るだけでなく、 夜 day 時点で alive な seat に絞る.
      for ( const [seat, status] of vs.statuses.entries() ) {
        if ( !isAliveAtNight(status, day) ) continue
        if (
          guardRoles.some(r => context.possibilities.hasRole(seat, r))
          || context.possibilities.hasRole(seat, foxRole)
        ) {
          continue DAY
        }
      }
    }
    return false
  }
  return true
}

// 変更あり版: 夜死者数の矛盾チェック + requireOneOf 制約の追加
export function updateDeathCountConstraints(
  context: AnalyzeContext,
  vs: VillageStatus,
  nightKillsByDay: Map<Day, Seat[]>,
  setup: Map<SystemRole, number>,
): boolean {
  const immoralists = countByTraitIn(setup, 'reactive', 'follow-fox-death')
  const guardRoles = rolesWithTraitIn(setup, 'action', 'guard')
  const guardCount = countByTraitIn(setup, 'action', 'guard')
  const followFoxRoles = rolesWithTraitIn(setup, 'reactive', 'follow-fox-death')
  DAY:
  for ( const [day, killed] of nightKillsByDay.entries() ) {
    if ( vs.day <= day ) continue DAY
    const addCount = context.deathChronicle.add[day]
    let expected = 1
    if ( addCount ) expected += addCount
    const actual = killed.length
    if ( actual === expected ) continue DAY
    if ( expected + immoralists < actual ) {
      return false
    }
    else if ( actual < expected - 1 ) {
      return false
    }
    else if ( expected < actual && actual <= expected + immoralists ) {
      const hamsterDiedThisNight = context.hamstersKilledByDivine.some(h => h.day === day)
      if ( hamsterDiedThisNight ) {
        for ( let i=0; i<immoralists; i++ ) {
          context.requireOneOf.push(
            killed.flatMap(seat => followFoxRoles.map(role => ({ seat, role })))
          )
        }
        continue DAY
      }
    }
    if ( actual < expected ) {
      // peace night (actual < expected) を成立させる説明:
      //   (A) 夜 day に alive な妖狐がいる (狼襲撃先 = 妖狐 → 襲撃免疫で死体なし)
      //   (B) 夜 day に alive な狩人がいる (護衛成功)
      // どちらの可能性も無ければ世界棄却. 片方しか可能性が無い場合, かつ setup の該当 role が
      // 1 体しかいない場合に限り, 「その 1 体は alive 側にいる」 と確定するので, 夜 day に
      // alive でない seat から該当 role を deny する. setup に 2 体以上いる場合は片方が
      // 死んでいても他方が alive なら peace を説明できるため deny できない (例: 狐 2 体構成で
      // 片方が呪殺後、 もう片方が襲撃先になる).
      let aliveFoxExists = false
      let aliveGuardExists = false
      for ( const [seat, status] of vs.statuses.entries() ) {
        if ( !isAliveAtNight(status, day) ) continue
        if ( context.possibilities.hasRole(seat, foxRole) ) aliveFoxExists = true
        if ( guardRoles.some(r => context.possibilities.hasRole(seat, r)) ) aliveGuardExists = true
      }
      if ( !aliveFoxExists && !aliveGuardExists ) return false
      if ( !aliveFoxExists && guardCount === 1 ) {
        // 説明は (B) のみ + 狩人は 1 体 → 夜 day に alive でない seat の guard 役職を deny
        for ( const [seat, status] of vs.statuses.entries() ) {
          if ( isAliveAtNight(status, day) ) continue
          for ( const r of guardRoles ) {
            if ( !context.possibilities.denyRole(seat, r) ) return false
          }
        }
      }
      if ( !aliveGuardExists && (setup.get(foxRole) ?? 0) === 1 ) {
        // 説明は (A) のみ + 狐は 1 体 → 夜 day に alive でない seat の foxRole を deny
        for ( const [seat, status] of vs.statuses.entries() ) {
          if ( isAliveAtNight(status, day) ) continue
          if ( !context.possibilities.denyRole(seat, foxRole) ) return false
        }
      }
      continue DAY
    }
    return false
  }
  return true
}

/**
 * action:divine trait 集約による狐呪殺の説明可能性チェック (読み取り専用).
 *
 * verifyDivineAbility が個別 role ごとに溜めた divineAliveMaxDay / divineTargetsByDay を
 * 使って、 「狐死日に占い能力者のいずれかが生きていた + その日の対象集合に狐 seat (または
 * 'unknown') が含まれる」を判定する.
 *
 * paparazzi 等の untrusted divine role がいる setup では、 seer 単独では説明できなくても
 * paparazzi が説明する可能性があるためここで集約判定する.
 */
export function checkDivineCoverage(context: AnalyzeContext): boolean {
  if (context.hamstersKilledByDivine.length === 0) return true

  // 1. 占い能力者の最大生存日 >= 狐呪殺最終日
  if (context.needDivineAliveAtDay != null
      && context.divineAliveMaxDay < context.needDivineAliveAtDay) {
    return false
  }

  // 2. 各狐呪殺について、 その日の divine target 集合に対象 seat (または 'unknown') を含む
  for (const { day, seat } of context.hamstersKilledByDivine) {
    const targets = context.divineTargetsByDay.get(day)
    if (!targets) return false
    if (!targets.has(seat) && !targets.has('unknown')) return false
  }

  return true
}

export function createDebugStash(): DebugStash {
  const roleTests = Object.fromEntries(
    [...systemRoles.keys()].map(r => [r, 0]),
  ) as Record<SystemRole, number>
  const roleTestPasses = Object.fromEntries(
    [...systemRoles.keys()].map(r => [r, 0]),
  ) as Record<SystemRole, number>
  return {
    finalizerRuns: 0, finalizerMiddle: 0, finalizerPasses: 0, finalizerFails: 0,
    roleTests, roleTestPasses,
    preFinalizeTests: 0, preFinalizePasses: 0,
  }
}

export type DebugStash = {
  finalizerRuns: number
  finalizerMiddle: number
  finalizerPasses: number
  finalizerFails: number
  roleTests: Record<SystemRole, number>
  roleTestPasses: Record<SystemRole, number>
  preFinalizeTests: number
  preFinalizePasses: number
}

export function finalize(
  context: AnalyzeContext,
  vs: VillageStatus,
  setup: Map<SystemRole, number>,
  conclusions: Possibilities,
  debugStash: DebugStash,
  hamsterWinPath: 'village' | 'wolf' | undefined,
  cachedSurvivors: Seat[],
  cachedSurvivingMap: Map<Seat, boolean>,
): void {
  debugStash.finalizerRuns++
  // 全 divine role の集約済み状態で狐呪殺の説明可能性を最終判定
  if (!checkDivineCoverage(context)) {
    debugStash.finalizerFails++
    return
  }
  dumpFinalizePre(context.possibilities)
  // ここまで処理が終わったところで、襲撃死した人物は非狼とみなす
  for ( const [seat, status] of vs.statuses.entries() ) {
    if ( !status.surviving && status.causeOfDeath === 'night_kill' ) {
      if ( context.possibilities.isFixed(seat) ) {
        continue
      }
      if (!context.possibilities.markAsHuman(seat)) {
        return
      }
    }
  }

  if (!context.possibilities.propagateFull()) {
    return
  }
  debugStash.finalizerMiddle++

  /*
  TODO: ここで、同じ役職の組み合わせをまとめてテストを行う
  const set = {}
  for ( const [seat, possibilities] of remained.entries() ) {
    const stringOfRoles = Array.from(possibilities).sort().join(',')
    set[stringOfRoles] ??= []
    set[stringOfRoles].push(seat)
  }
  */


  const survivors = cachedSurvivors
  // 狐陣営勝利の生存カウントは passive:fox-win-counter trait を持つ全役職 (妖狐 + 子狐) を対象にする.
  // die-when-divined (= 妖狐のみ) で数えると子狐生存だけで勝った世界線が拾えない.
  const hamsterWinRoles = rolesWithTraitIn(setup, 'passive', 'fox-win-counter')
  const setupFoxes = countByTraitIn(setup, 'passive', 'fox-win-counter')
  // 飽和ライン (maxSurvivingWolves) 算出時の hamster 数は候補ベース (hasRole) で setup 枠キャップ.
  // 確定数 (isActualRole) で数えると planBuilder が妖狐しか枝刈り固定しないので子狐は未確定の
  // まま 0 計上され、 「妖狐+子狐 両生存 + 狼 2 + 人間 2 = 飽和」 のような解を取り逃す.
  const possibleSurvivingHamsters = Math.min(
    setupFoxes,
    survivors.filter(seat =>
      hamsterWinRoles.some(r => context.possibilities.hasRole(seat, r))
    ).length
  )
  const setupWolves = countBySeerResultIn(setup, 'wolf')
  const maxSurvivingWolves = Math.min(
    setupWolves || Infinity,
    Math.floor((survivors.length - possibleSurvivingHamsters - 0.1) / 2)
  )

  const survivingMap = cachedSurvivingMap
  const condition = {
    minSurvivingWolves: 1,
    maxSurvivingWolves,
    minSurvivingHamsters: 0,
    maxSurvivingHamsters: setupFoxes,
  }

  // 村勝ちまたは狼勝ちの場合は、狼の生存数の条件を変更する
  if ( vs.result === 'werewolf_won' ) {
    condition.minSurvivingWolves = maxSurvivingWolves + 1
    condition.maxSurvivingWolves = Infinity
    condition.minSurvivingHamsters = 0
    condition.maxSurvivingHamsters = 0
  }
  else if ( vs.result === 'villager_won' ) {
    condition.minSurvivingWolves = 0
    condition.maxSurvivingWolves = 0
    condition.minSurvivingHamsters = 0
    condition.maxSurvivingHamsters = 0
  }

  // All seats fully determined → validate survival counts and skip solver
  let allFixed = true
  for (let i = 1; i < context.possibilities.possibilities.length; i++) {
    if (!context.possibilities.isFixed(i)) { allFixed = false; break }
  }
  if (allFixed) {
    // Count surviving wolves and hamsters
    let survWolves = 0, survHamsters = 0
    for (const seat of survivors) {
      if (context.possibilities.isActualRole(seat, wolfRole)) survWolves++
      if (hamsterWinRoles.some(r => context.possibilities.isActualRole(seat, r))) survHamsters++
    }
    const checkCondition = (minW: number, maxW: number, minH: number, maxH: number) =>
      survWolves >= minW && survWolves <= maxW && survHamsters >= minH && survHamsters <= maxH

    if (vs.result === 'werehamster_won') {
      if (hamsterWinPath !== 'wolf' && checkCondition(0, 0, 1, Infinity)) {
        debugStash.finalizerPasses++
        conclusions.union(context.possibilities)
      }
      if (hamsterWinPath !== 'village' && checkCondition(maxSurvivingWolves + 1, Infinity, 1, Infinity)) {
        debugStash.finalizerPasses++
        conclusions.union(context.possibilities)
      }
    } else if (checkCondition(condition.minSurvivingWolves, condition.maxSurvivingWolves, condition.minSurvivingHamsters, condition.maxSurvivingHamsters)) {
      debugStash.finalizerPasses++
      conclusions.union(context.possibilities)
    } else {
      debugStash.finalizerFails++
    }
    return
  }

  // 狐勝ちの場合だけは、狼全滅と飽和の両方を検証する
  // hamsterWinPath が指定されている場合は該当パスのみ実行
  if ( vs.result === 'werehamster_won') {
    if (hamsterWinPath !== 'wolf') {
      const conclusion = solvePossibilities(
        context.possibilities,
        survivingMap,
        0,
        0,
        1,
        Infinity,
        setup
      )
      if (conclusion) {
        debugStash.finalizerPasses++
        conclusions.union(conclusion)
      }
    }
    if (hamsterWinPath !== 'village') {
      const conclusion2 = solvePossibilities(
        context.possibilities,
        survivingMap,
        maxSurvivingWolves + 1,
        Infinity,
        1,
        Infinity,
        setup
      )
      if (conclusion2) {
        debugStash.finalizerPasses++
        conclusions.union(conclusion2)
      }
    }
  }
  else {
    const conclusion = solvePossibilities(
      context.possibilities,
      survivingMap,
      condition.minSurvivingWolves,
      condition.maxSurvivingWolves,
      condition.minSurvivingHamsters,
      condition.maxSurvivingHamsters,
      setup
    )
    if ( !conclusion ) {
      debugStash.finalizerFails++
      return
    }
    debugStash.finalizerPasses++
    conclusions.union(conclusion)
  }
}

File: src/retar-rs/src/finalizer.rs
use crate::types::{CauseOfDeath, EnumSpecies, RoleTrait, SeatStatus, VillageStatus, VillageResult, SystemRole, Seat, Day};
use crate::possibilities::Possibilities;
use crate::role_testers::{AnalyzeContext, DivineTarget};
use crate::role_sets::{
    count_by_seer_result_in, count_by_trait_in, roles_by_seer_result, roles_with_trait_in,
};
use crate::solver::solve_possibilities;
use std::collections::BTreeMap;
use std::sync::LazyLock;

// hot path で繰り返し参照するため module-level lazy 解決. TS finalizer.ts:14-17 と同じ pattern.
static WOLF_ROLES: LazyLock<Vec<SystemRole>> = LazyLock::new(|| roles_by_seer_result(EnumSpecies::Wolf));

/// seat が夜 `day` の時点で行動可能 (alive) かどうか.
/// 夜 D の cause_of_death = NightKill は「その夜に襲撃で死亡」 = その夜の行動は可能, とみなす
/// (verify_divine_ability の seer max_active_day と同じ流儀).
/// Execution は昼の処刑なので died_day == D でも夜 D は alive ではない.
fn is_alive_at_night(status: &SeatStatus, day: Day) -> bool {
    if status.surviving {
        return true;
    }
    let died = status.died_day.unwrap_or(0);
    if died > day {
        return true;
    }
    if died == day && status.cause_of_death == CauseOfDeath::NightKill {
        return true;
    }
    false
}

const ROLE_COUNT: usize = SystemRole::ALL.len();

#[derive(Debug, Clone)]
pub struct DebugStash {
    pub finalizer_runs: u32,
    pub finalizer_middle: u32,
    pub finalizer_passes: u32,
    pub finalizer_fails: u32,
    pub role_tests: [u32; ROLE_COUNT],
    pub role_test_passes: [u32; ROLE_COUNT],
    pub pre_finalize_tests: u32,
    pub pre_finalize_passes: u32,
}

impl Default for DebugStash {
    fn default() -> Self {
        Self {
            finalizer_runs: 0,
            finalizer_middle: 0,
            finalizer_passes: 0,
            finalizer_fails: 0,
            role_tests: [0; ROLE_COUNT],
            role_test_passes: [0; ROLE_COUNT],
            pre_finalize_tests: 0,
            pre_finalize_passes: 0,
        }
    }
}

pub fn create_debug_stash() -> DebugStash {
    DebugStash::default()
}

/// action:divine trait 集約による狐呪殺の説明可能性チェック (読み取り専用).
///
/// verify_divine_ability が個別 role ごとに溜めた divine_alive_max_day / divine_targets_by_day を
/// 使って、 「狐死日に占い能力者のいずれかが生きていた + その日の対象集合に狐 seat (または
/// Unknown) が含まれる」を判定する.
///
/// paparazzi 等の untrusted divine role がいる setup では、 seer 単独では説明できなくても
/// paparazzi が説明する可能性があるためここで集約判定する.
pub fn check_divine_coverage(context: &AnalyzeContext) -> bool {
    if context.hamsters_killed_by_divine.is_empty() {
        return true;
    }

    // 1. 占い能力者の最大生存日 >= 狐呪殺最終日
    if let Some(need_day) = context.need_divine_alive_at_day {
        if context.divine_alive_max_day < need_day {
            return false;
        }
    }

    // 2. 各狐呪殺について、 その日の divine target 集合に対象 seat (または Unknown) を含む
    for hk in &context.hamsters_killed_by_divine {
        let targets = match context.divine_targets_by_day.get(&hk.day) {
            None => return false,
            Some(t) => t,
        };
        if !targets.contains(&DivineTarget::Seat(hk.seat))
            && !targets.contains(&DivineTarget::Unknown)
        {
            return false;
        }
    }

    true
}

/// Hamster win path for 2-pass analysis
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum HamsterWinPath {
    Village,
    Wolf,
}

/// Validate night death counts against role hypotheses.
pub fn check_death_counts(
    context: &AnalyzeContext,
    vs: &VillageStatus,
    night_kills_by_day: &BTreeMap<Day, Vec<Seat>>,
    setup: &BTreeMap<SystemRole, u32>,
) -> bool {
    let guard_roles = roles_with_trait_in(setup, RoleTrait::ActionGuard);
    let fox_roles = roles_with_trait_in(setup, RoleTrait::PassiveDieWhenDivined);
    for (&day, killed) in night_kills_by_day {
        if vs.day <= day {
            continue;
        }
        let add_count = context.death_chronicle.add[day as usize];
        let mut expected: i32 = 1;
        if add_count != 0 {
            expected += add_count as i32;
        }
        let actual = killed.len() as i32;
        let immoralists = count_by_trait_in(setup, RoleTrait::ReactiveFollowFoxDeath) as i32;
        if actual == expected {
            continue;
        }
        if expected + immoralists < actual {
            return false;
        } else if actual < expected - 1 {
            return false;
        } else if expected < actual && actual <= expected + immoralists {
            let hamster_died_this_night = context
                .hamsters_killed_by_divine
                .iter()
                .any(|h| h.day == day);
            if hamster_died_this_night {
                // Would add requireOneOf constraints, but we're in an immutable check
                // The caller handles this in the mutable version
                continue;
            }
        }
        if actual < expected {
            // peace night (actual < expected) を成立させる説明:
            //   (A) 夜 day に alive な妖狐がいる (狼襲撃先 = 妖狐 → 襲撃免疫で死体なし)
            //   (B) 夜 day に alive な狩人がいる (護衛成功)
            let mut has_protector = false;
            for (&seat, status) in &vs.statuses {
                if !is_alive_at_night(status, day) {
                    continue;
                }
                if guard_roles.iter().any(|&r| context.possibilities.has_role(seat, r))
                    || fox_roles.iter().any(|&r| context.possibilities.has_role(seat, r))
                {
                    has_protector = true;
                    break;
                }
            }
            if has_protector {
                continue;
            }
        }
        return false;
    }
    true
}

/// Mutable version that can add requireOneOf constraints
pub fn update_death_count_constraints(
    context: &mut AnalyzeContext,
    vs: &VillageStatus,
    night_kills_by_day: &BTreeMap<Day, Vec<Seat>>,
    setup: &BTreeMap<SystemRole, u32>,
) -> bool {
    let guard_roles = roles_with_trait_in(setup, RoleTrait::ActionGuard);
    let guard_count = count_by_trait_in(setup, RoleTrait::ActionGuard);
    let follow_fox_roles = roles_with_trait_in(setup, RoleTrait::ReactiveFollowFoxDeath);
    let fox_roles = roles_with_trait_in(setup, RoleTrait::PassiveDieWhenDivined);
    let fox_count = count_by_trait_in(setup, RoleTrait::PassiveDieWhenDivined);
    for (&day, killed) in night_kills_by_day {
        if vs.day <= day {
            continue;
        }
        let add_count = context.death_chronicle.add[day as usize];
        let mut expected: i32 = 1;
        if add_count != 0 {
            expected += add_count as i32;
        }
        let actual = killed.len() as i32;
        let immoralists = count_by_trait_in(setup, RoleTrait::ReactiveFollowFoxDeath) as i32;
        if actual == expected {
            continue;
        }
        if expected + immoralists < actual {
            return false;
        } else if actual < expected - 1 {
            return false;
        } else if expected < actual && actual <= expected + immoralists {
            let hamster_died_this_night = context
                .hamsters_killed_by_divine
                .iter()
                .any(|h| h.day == day);
            if hamster_died_this_night {
                for _ in 0..immoralists {
                    context.require_one_of.push(
                        killed
                            .iter()
                            .flat_map(|&seat| {
                                follow_fox_roles.iter().map(move |&role| {
                                    crate::role_testers::SeatRole { seat, role }
                                })
                            })
                            .collect(),
                    );
                }
                continue;
            }
        }
        if actual < expected {
            // peace night (actual < expected) を成立させる説明:
            //   (A) 夜 day に alive な妖狐がいる (狼襲撃先 = 妖狐 → 襲撃免疫で死体なし)
            //   (B) 夜 day に alive な狩人がいる (護衛成功)
            // どちらの可能性も無ければ世界棄却. 片方しか可能性が無い場合, かつ setup の該当 role が
            // 1 体しかいない場合に限り, 「その 1 体は alive 側にいる」 と確定するので, 夜 day に
            // alive でない seat から該当 role を deny する. setup に 2 体以上いる場合は片方が
            // 死んでいても他方が alive なら peace を説明できるため deny できない (例: 狐 2 体構成で
            // 片方が呪殺後、 もう片方が襲撃先になる).
            let mut alive_fox_exists = false;
            let mut alive_guard_exists = false;
            for (&seat, status) in &vs.statuses {
                if !is_alive_at_night(status, day) {
                    continue;
                }
                if fox_roles.iter().any(|&r| context.possibilities.has_role(seat, r)) {
                    alive_fox_exists = true;
                }
                if guard_roles.iter().any(|&r| context.possibilities.has_role(seat, r)) {
                    alive_guard_exists = true;
                }
            }
            if !alive_fox_exists && !alive_guard_exists {
                return false;
            }
            if !alive_fox_exists && guard_count == 1 {
                // 説明は (B) のみ + 狩人は 1 体 → 夜 day に alive でない seat の guard 役職を deny
                let dead_seats: Vec<Seat> = vs
                    .statuses
                    .iter()
                    .filter(|(_, status)| !is_alive_at_night(status, day))
                    .map(|(&seat, _)| seat)
                    .collect();
                for seat in dead_seats {
                    for &r in &guard_roles {
                        if !context.possibilities.deny_role(seat, r) {
                            return false;
                        }
                    }
                }
            }
            if !alive_guard_exists && fox_count == 1 {
                // 説明は (A) のみ + 狐は 1 体 → 夜 day に alive でない seat の fox 役職を deny
                let dead_seats: Vec<Seat> = vs
                    .statuses
                    .iter()
                    .filter(|(_, status)| !is_alive_at_night(status, day))
                    .map(|(&seat, _)| seat)
                    .collect();
                for seat in dead_seats {
                    for &r in &fox_roles {
                        if !context.possibilities.deny_role(seat, r) {
                            return false;
                        }
                    }
                }
            }
            continue;
        }
        return false;
    }
    true
}

pub fn finalize(
    context: &mut AnalyzeContext,
    vs: &VillageStatus,
    setup: &BTreeMap<SystemRole, u32>,
    conclusions: &mut Possibilities,
    debug_stash: &mut DebugStash,
    hamster_win_path: Option<HamsterWinPath>,
    cached_survivors: &[Seat],
    cached_surviving_map: &BTreeMap<Seat, bool>,
) {
    debug_stash.finalizer_runs += 1;
    // 全 divine role の集約済み状態で狐呪殺の説明可能性を最終判定
    if !check_divine_coverage(context) {
        debug_stash.finalizer_fails += 1;
        return;
    }
    #[cfg(feature = "dump")] crate::dump::finalize_pre(&context.possibilities);

    // Mark night-kill victims as non-wolf
    for (&seat, status) in &vs.statuses {
        if !status.surviving && status.cause_of_death == CauseOfDeath::NightKill {
            if context.possibilities.is_fixed(seat) {
                continue;
            }
            if !context.possibilities.mark_as_human(seat) {
                return;
            }
        }
    }

    if !context.possibilities.propagate_full() {
        return;
    }
    debug_stash.finalizer_middle += 1;

    let survivors = cached_survivors;
    // 狐陣営勝利の生存カウントは passive:fox-win-counter trait を持つ全役職 (妖狐 + 子狐) を対象にする.
    // die-when-divined (= 妖狐のみ) で数えると子狐生存だけで勝った世界線が拾えない.
    let hamster_win_roles = roles_with_trait_in(setup, RoleTrait::PassiveFoxWinCounter);
    let setup_foxes = count_by_trait_in(setup, RoleTrait::PassiveFoxWinCounter);
    // 飽和ライン (max_surviving_wolves) 算出時の hamster 数は候補ベース (has_role) で setup 枠キャップ.
    // 確定数 (is_actual_role) で数えると plan_builder が妖狐しか枝刈り固定しないので子狐は未確定の
    // まま 0 計上され、 「妖狐+子狐 両生存 + 狼 2 + 人間 2 = 飽和」 のような解を取り逃す.
    let possible_surviving_hamsters = setup_foxes.min(
        survivors
            .iter()
            .filter(|&&seat| {
                hamster_win_roles
                    .iter()
                    .any(|&r| context.possibilities.has_role(seat, r))
            })
            .count() as u32,
    );

    // TS finalizer.ts と同じ. setup に wolf 役職が無ければ 0 が返る.
    let wolf_count = count_by_seer_result_in(setup, EnumSpecies::Wolf);
    let raw_max = if survivors.len() as u32 > possible_surviving_hamsters {
        (survivors.len() as u32 - possible_surviving_hamsters - 1) / 2
    } else {
        0
    };
    let max_surviving_wolves = wolf_count.min(raw_max);

    let mut condition = SolveCondition {
        min_surviving_wolves: 1,
        max_surviving_wolves,
        min_surviving_hamsters: 0,
        max_surviving_hamsters: setup_foxes,
    };

    if vs.result == Some(VillageResult::WerewolfWon) {
        condition.min_surviving_wolves = max_surviving_wolves + 1;
        condition.max_surviving_wolves = u32::MAX;
        condition.min_surviving_hamsters = 0;
        condition.max_surviving_hamsters = 0;
    } else if vs.result == Some(VillageResult::VillagerWon) {
        condition.min_surviving_wolves = 0;
        condition.max_surviving_wolves = 0;
        condition.min_surviving_hamsters = 0;
        condition.max_surviving_hamsters = 0;
    }

    // All seats fully determined → validate survival counts directly
    let all_fixed = (1..context.possibilities.possibilities.len())
        .all(|i| context.possibilities.is_fixed(i as Seat));

    if all_fixed {
        let mut surv_wolves = 0u32;
        let mut surv_hamsters = 0u32;
        for &seat in survivors {
            if WOLF_ROLES.iter().any(|&r| context.possibilities.is_actual_role(seat, r)) {
                surv_wolves += 1;
            }
            if hamster_win_roles
                .iter()
                .any(|&r| context.possibilities.is_actual_role(seat, r))
            {
                surv_hamsters += 1;
            }
        }

        let check = |min_w, max_w, min_h, max_h| {
            surv_wolves >= min_w && surv_wolves <= max_w && surv_hamsters >= min_h && surv_hamsters <= max_h
        };

        if vs.result == Some(VillageResult::WerehamsterWon) {
            if hamster_win_path != Some(HamsterWinPath::Wolf) && check(0, 0, 1, u32::MAX) {
                debug_stash.finalizer_passes += 1;
                conclusions.union(&context.possibilities);
            }
            if hamster_win_path != Some(HamsterWinPath::Village)
                && check(max_surviving_wolves.saturating_add(1), u32::MAX, 1, u32::MAX)
            {
                debug_stash.finalizer_passes += 1;
                conclusions.union(&context.possibilities);
            }
        } else if check(
            condition.min_surviving_wolves,
            condition.max_surviving_wolves,
            condition.min_surviving_hamsters,
            condition.max_surviving_hamsters,
        ) {
            debug_stash.finalizer_passes += 1;
            conclusions.union(&context.possibilities);
        } else {
            debug_stash.finalizer_fails += 1;
        }
        return;
    }

    // Werehamster won: 2-pass
    if vs.result == Some(VillageResult::WerehamsterWon) {
        if hamster_win_path != Some(HamsterWinPath::Wolf) {
            if let Some(conclusion) = solve_possibilities(
                &context.possibilities,
                cached_surviving_map,
                0,
                0,
                1,
                u32::MAX,
                setup,
            ) {
                debug_stash.finalizer_passes += 1;
                conclusions.union(&conclusion);
            }
        }
        if hamster_win_path != Some(HamsterWinPath::Village) {
            if let Some(conclusion) = solve_possibilities(
                &context.possibilities,
                cached_surviving_map,
                max_surviving_wolves + 1,
                u32::MAX,
                1,
                u32::MAX,
                setup,
            ) {
                debug_stash.finalizer_passes += 1;
                conclusions.union(&conclusion);
            }
        }
    } else {
        let conclusion = solve_possibilities(
            &context.possibilities,
            cached_surviving_map,
            condition.min_surviving_wolves,
            condition.max_surviving_wolves,
            condition.min_surviving_hamsters,
            condition.max_surviving_hamsters,
            setup,
        );
        match conclusion {
            None => {
                debug_stash.finalizer_fails += 1;
            }
            Some(c) => {
                debug_stash.finalizer_passes += 1;
                conclusions.union(&c);
            }
        }
    }
}

struct SolveCondition {
    min_surviving_wolves: u32,
    max_surviving_wolves: u32,
    min_surviving_hamsters: u32,
    max_surviving_hamsters: u32,
}
